unarchive: Various changes/improvements, add debugging output

dagwieers opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the plugin/module/task -->
unarchive

##### ANSIBLE VERSION
<!---
Paste verbatim output from “ansible --version” between quotes below,
this is to help the Ansible team determine if this is a version specific
issue which is being fixed.
-->
v2.2

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnnn" in your commit
message and your description; but you should still explain what the change does.
-->
Various changes and improvements:

- Fixed a few issues with the existing unzip support which were related to Zipfile (which is fundamentally incompatible with unzip excludes/opts).  And issues related to forced ownership.

- Improvements to the UnarchiveError Exception, so it can report more detailed information to why an archive cannot be handled.

- A bugfix relatd to unicode filenames and crc32 handling.

- Renamed a few things so the code is easier to read.

- Ensure gtar considers size changes to unarchive its content. This fixes ansible/ansible-modules-core#5149

- Added a note about ZIP archives without directories. This fixes ansible/ansible-modules-core#5744.

The unarchive module is a hard beast to tame, corner-cases and platform-specific issues make it hard to troubleshoot individual issues.

- Improves the default output (e.g. unzip no longer shows standard output, just like gtar) and adds the ability to get debug-output when using -vvvv (verbosity >= 4) or -vvvvv. So that when issues are i reported we have the means to look what is going on internally.

- Because the output by default is reduced, proper warnings/errors are now more obvious to end-users, e.g.

```
TASK [unarchive]
***************************************************************
task path: /home/dag/home-made/ansible.testing/test126.yml:11
changed: [localhost] => {"changed": true, "dest": "/tmp/test126a", "extract_results": {"cmd":
["/usr/bin/unzip", "-q", "-o", "/home/dag/.ansible/tmp/ansible-tmp-1479989967.66-3171393713112/source",
"-x", "some.directory/", "another.directory/simple.test.file.txt", "-d", "/tmp/test126a"], "err":
"caution: excluded filename not matched: some.directory/\ncaution: excluded filename not matched:
another.directory/simple.test.file.txt\n", "out": "", "rc": 0}, "gid": 500, "group": "dag", "handler":
"ZipArchive", "mode": "0775", "owner": "dag", "secontext": "unconfined_u:object_r:user_tmp_t:s0",
"size": 4096, "src": "/home/dag/.ansible/tmp/ansible-tmp-1479989967.66-3171393713112/source",
"state": "directory", "uid": 500}
```
Whereas before the complete list of deflated files was being shown (when using -v) which was much more than anyone was willing to look at.

*Migrated from https://github.com/ansible/ansible-modules-core/pull/5721*

dagwieers commented:
Migrated from https://github.com/ansible/ansible-modules-core/pull/5721

mattclay commented:
CI failure on FreeBSD due to:

```
2017-02-14 12:02:54 TASK [unarchive : unarchive a zip file] ****************************************
2017-02-14 12:02:54 fatal: [testhost]: FAILED! => {"changed": false, "dest": "/root/ansible_testing/test-unarchive-zip", "details": "[Errno 2] No such file or directory: '/root/ansible_testing/test-unarchive-zip/foo-unarchive.txt'", "failed": true, "gid": 0, "group": "wheel", "handler": "ZipArchive", "mode": "0755", "msg": "Unexpected error when accessing exploded file", "owner": "root", "size": 512, "src": "/root/ansible_testing/test-unarchive.zip", "state": "directory", "uid": 0}
```

dagwieers commented:
No CI, close and re-open.

mattclay commented:
CI failure on FreeBSD. The error is fairly lengthy, so I won't copy it here.

dagwieers commented:
The problem is known here, FreeBSD unzip is quite a different beast than all the other OSes. Much like bsdtar vs gtar :-/ The solution is not yet known as it is harder to debug.

calfonso commented:
@dagwieers could you rebase this and fix up the BSD zip case?

dagwieers commented:
@calfonso Right, that was the plan. But fixing BSD was much harder than expected. Feel free to create a PR for that one against my branch.

mattclay commented:
CI failure in integration tests due to traceback:

```
Traceback (most recent call last):
  File "/tmp/ansible_OCWhAc/ansible_module_unarchive.py", line 867, in <module>
    main()
  File "/tmp/ansible_OCWhAc/ansible_module_unarchive.py", line 817, in main
    handler = pick_handler(src, dest, file_args, module)
  File "/tmp/ansible_OCWhAc/ansible_module_unarchive.py", line 739, in pick_handler
    (can_handle, info) = obj.can_handle_archive()
  File "/tmp/ansible_OCWhAc/ansible_module_unarchive.py", line 558, in can_handle_archive
    e = get_exception()
NameError: global name 'get_exception' is not defined
```

mattclay commented:
CI failure in FreeBSD integration tests:

https://app.shippable.com/github/ansible/ansible/runs/32038/8/tests

dagwieers commented:
Yup, we know. Still the same thing: https://github.com/ansible/ansible/pull/19015#issuecomment-281199523

dagwieers commented:
So I am very happy to drop FeeBSD support altogether. If they want to have a freaky unzip with different behavior, this module is not for them.

Maybe it helps to get the pure-python implementation finally going.

dagwieers committed:
unarchive: Various improvements, adds debug output

Various changes and improvements:

- Fixed a few issues with the existing unzip support which were related
  to Zipfile (which is fundamentally incompatible with unzip
excludes/opts).  And issues related to forced ownership.

- Improvements to the UnarchiveError Exception, so it can report more
  detailed information to why an archive cannot be handled.

- A bugfix relatd to unicode filenames and crc32 handling.

- Renamed a few things so the code is easier to read.

- Ensure gtar considers size changes to unarchive its content. This
  fixes ansible/ansible-modules-core#5149

- Added a note about ZIP archives without directories. This fixes
  ansible/ansible-modules-core#5744.

The unarchive module is a hard beast to tame, corner-cases and
platform-specific issues make it hard to troubleshoot individual issues.

- Improves the default output (e.g. unzip no longer shows standard
  output, just like gtar) and adds the ability to get debug-output when
using -vvvv (verbosity >= 4) or -vvvvv. So that when issues are i
reported we have the means to look what is going on internally.

- Because the output by default is reduced, proper warnings/errors are
  now more obvious to end-users

jborean93 commented:
Closing due to the age of the PR.
